staking/slash/double-sign.go

```diff
@@ -1,52 +1,385 @@
-func Verify(candidate *Record) error {
+func Verify(chain CommitteeReader, candidate *Record) error {
+	first, second := candidate.AlreadyCastBallot, candidate.DoubleSignedBallot
+	if shard.CompareBlsPublicKey(first.SignerPubKey, second.SignerPubKey) != 0 {
+		k1, k2 := first.SignerPubKey.Hex(), second.SignerPubKey.Hex()
+		return errors.Wrapf(
+			errBLSKeysNotEqual, "%s %s", k1, k2,
+		)
+	}
+	superCommittee, err := chain.ReadShardState(candidate.Evidence.Epoch)
+
+	if err != nil {
+		return err
+	}
+
+	subCommittee := superCommittee.FindCommitteeByID(
+		candidate.Evidence.ShardID,
+	)
+
+	if subCommittee == nil {
+		return errors.Wrapf(
+			errShardIDNotKnown, "given shardID %d", candidate.Evidence.ShardID,
+		)
+	}
+
+	if _, err := subCommittee.AddressForBLSKey(second.SignerPubKey); err != nil {
+		return err
+	}
+	// candidate.ConflictingBallots
+
+	// TODO Why this one printng have 00000000 for signature? something wrong earlier
+	// TODO need to finish this implementation
+
 	return nil
 }
 
-// Apply ..
-func Apply(state *state.DB, slashes []byte) error {
+var (
+	errBLSKeysNotEqual = errors.New(
+		"bls keys in ballots accompanying slash evidence not equal ",
+	)
+	errSlashDebtNotFullyAccountedFor = errors.New(
+		"slash debt was not fully accounted for, still non-zero",
+	)
+	errShardIDNotKnown              = errors.New("nil subcommittee for shardID")
+	errValidatorNotFoundDuringSlash = errors.New("validator not found")
+	zero                            = numeric.ZeroDec()
+	oneDoubleSignerRate             = numeric.MustNewDecFromStr("0.02")
+	oneHundredPercent               = numeric.NewDec(1)
+	fiftyPercent                    = numeric.MustNewDecFromStr("0.5")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(amount *big.Int, rate numeric.Dec) *big.Int {
+	return numeric.NewDecFromBigInt(
+		amount,
+	).Mul(rate).TruncateInt()
+}
+
+func delegatorSlashApply(
+	snapshot, current *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for i, delegationSnapshot := range snapshot.Delegations {
+		slashDebt := applySlashRate(delegationSnapshot.Amount, rate)
+		halfOfSlashDebt := new(big.Int).Div(slashDebt, common.Big2)
+		snapshotAddr := delegationSnapshot.DelegatorAddress
+		for j, delegationNow := range current.Delegations {
+			if nowAmt := delegationNow.Amount; delegationNow.DelegatorAddress == snapshotAddr {
+				state.AddBalance(reporter, halfOfSlashDebt)
+				// NOTE only need to pay snitch here
+				slashTrack.TotalSnitchReward.Add(
+					slashTrack.TotalSnitchReward, halfOfSlashDebt,
+				)
+				// NOTE handle validator self delegation special case
+				if i == validatorsOwnDel && j == validatorsOwnDel {
+					// say my slash debt is 1.6, and my current amount is 1.2
+					// then while I can't drop below 1, I could still pay 0.2
```

Comment:
We should just change the sanityCheck to take into account the Banned field. Having complicated logic here to stay above 1 or minSelfDelegation is not necessary.